### PR TITLE
Initial version

### PR DESCRIPTION
**Pull request by michal-matusik90**

### CODE REVIEW

File: CodeGenerators/ICodeGenerator.cs [hunk 1]
```diff
@@ -0,0 +1,9 @@
+﻿using EntityManager.ViewModels;
```
Comment (Muzlu) [inline CodeGenerators/ICodeGenerator.cs]: moge nawet dawac komenty

**Review by RoboCt — COMMENTED** [review]
asdasdsadsa

File: CodeGenerators/ICodeGenerator.cs [hunk 2]
```diff
@@ -0,0 +1,9 @@
...
+
+namespace EntityManager.CodeGenerators
+{
+    public interface ICodeGenerator
```
Comment (Muzlu) [inline CodeGenerators/ICodeGenerator.cs]: test komenta

File: StringExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    public static class StringExtensions
+    {
+        public static string FirstCharToUpper(this string input)
+        {
+            if (!string.IsNullOrEmpty(input))
+            {
+                return input.First().ToString().ToUpper() + input.Substring(1);
+            }
+            return string.Empty;
+        }
```
Comment (RoboCt) [inline StringExtensions.cs:15]: this function shouldnt work on empty string -> throw error/exception<br>
Maybe you should use ToCharArray(); method?<br>
var bla = input.ToCharArray();<br>
bla[0] = char.ToUpper(bla[0]);<br>
return bla.ToString(); // or return new string(bla); <br>
both are good -> however char method is much faster. matter of choice / performance

File: StringExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+        public static string FirstCharToLower(this string input)
```
Comment (RoboCt) [inline StringExtensions.cs:17]: same as above but with char.ToLower(bla);

File: StringExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+            }
+            return string.Empty;
+        }
+
+        public static string DropSuffix(this string input, string suffix)
+        {
+            if (input.EndsWith(suffix))
+            {
+                return input.Remove(input.Length - suffix.Length);
+            }
```
Comment (RoboCt) [inline StringExtensions.cs:31]: you can remove brackets in 1 line if's<br>
            if (input.EndsWith(suffix))<br>
            return input.Remove(input.Length - suffix.Length);<br>

File: StringExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+            {
+                return input;
+            }
+
+            if (!input.Contains("_"))
+            {
+                return input;
+            }
+
+            string[] words = input.Split("_", StringSplitOptions.RemoveEmptyEntries);
```
Comment (RoboCt) [inline StringExtensions.cs:52]: **var** words = input.Split("_", StringSplitOptions.RemoveEmptyEntries);<hr>
everywhere where possible

File: StringExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+                result += word.Substring(0, 1).ToUpper() + word.Substring(1);
+            }
+
+            return result;
+        }
+
+        public static string Indent(this string input, int level = 0)
+        {
+            return input.Insert(0,new string(' ', level * 4));
+        }
```
Comment (RoboCt) [inline StringExtensions.cs:66]: In functions with only return statement -> remove brackets and use c# 7 => notation.
`public static string Indent(this string input, int level = 0)
      =>  input.Insert(0,new string(' ', level * 4));`

File: StringExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+            {
+                return input;
+            }
+
+            string[] words = input.Split("_", StringSplitOptions.RemoveEmptyEntries);
+            string result = string.Empty;
+
+            foreach (string word in words)
+            {
+                result += word.Substring(0, 1).ToUpper() + word.Substring(1);
```
Comment (RoboCt) [inline StringExtensions.cs:57]: if there is many strings you are adding togather or make operations -> use string builder<br>
this is good to if u do not making much calculations etc

File: CodeGenerators/ICodeGenerator.cs [hunk 9]
```diff
@@ -0,0 +1,9 @@
...
+    {
+        string Generate(ShellViewModel viewModel);
+    }
+}
```
Comment (RoboCt) [inline CodeGenerators/ICodeGenerator.cs]: no new line on eof

File: CodeGenerators/SqlCodeGenerator.cs [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+                $"CREATE TABLE { viewModel.TableName } (",
+                string.Join(Environment.NewLine, joinedColumns),
+                ");",
+                string.Empty
+            };
+            
+            return string.Join(Environment.NewLine, codeBlock);
+        }
+    }
+}
```
Comment (RoboCt) [inline CodeGenerators/SqlCodeGenerator.cs]: eof

File: Models/ObjectsFactoryModel.cs [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+namespace EntityManager.Models
+{
+    public class ObjectsFactoryModel
+    {
+        public string VbName { get; set; }
+        public string Description { get; set; }
+        public List<FunctionModel> Functions { get; set; } = new List<FunctionModel>();
+
+        public int GetMaxFunctionNameLenght() => Functions.Max(f => f.VbaName.Length);
+        public int GetMaxFunctionDeclarationLenght() => Functions.Max(f => f.VbaDeclaration.Length);
```
Comment (RoboCt) [inline Models/ObjectsFactoryModel.cs:14]: maybe these could be only getters? without setter<br>
or private variables -> and these get updated everytime item is added to collection -> in addFunction function.<br> with max -> everytime we use linq to go though whole collection -> with private variable we will just compare lenght of 2 variables.

File: Readers/ObjectsFactoryReader.cs [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+
+        private void ParseFunctions()
+        {
+            var regexPattern = @"^Public Function (?<functionName>\w+)\(\) As AllInterfaces.FactorableInterface: *Set \1 = New (?<functionType>\w+\.\w+): *End Function";
+            var regex = new Regex(regexPattern, RegexOptions.Multiline);
+            var matches = regex.Matches(fileContent);
+
+            foreach (Match match in matches)
+            {
+                objectFactory.Functions.Add(new FunctionModel(match.Groups["functionType"].Value));
```
Comment (RoboCt) [inline Readers/ObjectsFactoryReader.cs:50]: you have function `AddFunction` in objectfactorymodel but u use Functions property directly.<br>
to collections items should be added by using function or custom setter (both with validation what we are adding)

File: Models/ObjectsFactoryModel.cs [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+
+        public ObjectsFactoryModel AddFunction(string functionType)
+        {
+            Functions.Add(new FunctionModel(functionType));
+            return this;
```
Comment (RoboCt) [inline Models/ObjectsFactoryModel.cs:19]: why this model is returned? there is nothing to chain?

File: Models/ObjectsFactoryModel.cs [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+using System.Collections.Generic;
+using System.Linq;
+
+namespace EntityManager.Models
+{
+    public class ObjectsFactoryModel
+    {
+        public string VbName { get; set; }
+        public string Description { get; set; }
+        public List<FunctionModel> Functions { get; set; } = new List<FunctionModel>();
```
Comment (RoboCt) [inline Models/ObjectsFactoryModel.cs:11]: maybe you should use IEnumerable<T> or ICollection<T> -> you do not need whole list implementation.<br>
you are using .Add function below then ICollection would be better -> all depends on usage needed.<br>
`public ICollection<FunctionModel> Functions { get; set; } = new List<FunctionModel>();`

File: Writers/SqlWriter.cs [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+        }
+
+        public void Write(ShellViewModel model)
+        {
+            string fileContents;
+            using (StreamReader reader = new StreamReader(SqlFilePath))
+            {
+                fileContents = reader.ReadToEnd();
+                reader.Close();
+            }
```
Comment (RoboCt) [inline Writers/SqlWriter.cs:25]: if you use File.WriteAllText at the and, is it not better to use File.ReadAllText in here?

File: Writers/SqlWriter.cs [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+            var tableExists = regex.IsMatch(fileContents);
+
+            if (tableExists)
+            {
+                fileContents = regex.Replace(fileContents, sqlCode);
+            }
+            else
+            {
+                fileContents += sqlCode;
+            }
```
Comment (RoboCt) [inline Writers/SqlWriter.cs:41]: remove brackets

File: StringExtensions.cs (hunk 3 above)
Comment (michal-matusik90) [reply StringExtensions.cs:15]: Changed to char.ToUpper() and throwing error if string is null or empty

File: StringExtensions.cs (hunk 4 above)
Comment (michal-matusik90) [reply StringExtensions.cs:17]: Changed as above

File: StringExtensions.cs (hunk 5 above)
Comment (michal-matusik90) [reply StringExtensions.cs:31]: Tim Corey advice was not to do so.

File: StringExtensions.cs (hunk 6 above)
Comment (michal-matusik90) [reply StringExtensions.cs:52]: Changed

File: CodeGenerators/ICodeGenerator.cs (hunk 9 above)
Comment (michal-matusik90) [reply CodeGenerators/ICodeGenerator.cs]: Changed globally

File: Models/ObjectsFactoryModel.cs (hunk 14 above)
Comment (michal-matusik90) [reply Models/ObjectsFactoryModel.cs:11]: Changed to ICollection

File: Writers/SqlWriter.cs (hunk 15 above)
Comment (michal-matusik90) [reply Writers/SqlWriter.cs:25]: Caused errors of file being busy when at WriteAllText line

File: Models/ObjectsFactoryModel.cs (hunk 11 above)
Comment (michal-matusik90) [reply Models/ObjectsFactoryModel.cs:14]: it is used only once in whole app, doesnt really matter much.

File: Readers/ObjectsFactoryReader.cs (hunk 12 above)
Comment (michal-matusik90) [reply Readers/ObjectsFactoryReader.cs:50]: Changed to use AddFunction of ObjectFactoryModel

**Comment by michal-matusik90** [comment]
String builder to be implemented globally. Readers and Writers to be refactored.

**Pull request merged**